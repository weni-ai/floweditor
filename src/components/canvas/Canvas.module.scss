@import 'variables.module.scss';
@import '~@weni/unnnic-system/src/assets/scss/unnnic.scss';

$background_offset: ($grid_size / 2) + ($node_padding) + px;

.canvas_container {
  overflow: hidden;
  width: 100%;
  margin-top: -281px;
}

.canvas_background {
  background-color: $unnnic-color-background-snow;
  background-position: 0p x 0px;
  background-image: radial-gradient(
    circle at center,
    $unnnic-color-neutral-soft 0,
    $unnnic-color-neutral-soft 1px,
    $unnnic-color-background-snow 2px,
    $unnnic-color-background-snow 100%);
  background-size: 40px 40px;
}
.canvas {
  position: relative;
  height: 100vh;
  width: 100%;
  overflow: hidden;
  outline: none;
}

.canvas:after {
  content: '';
  position: absolute;
  top: -134px;
  bottom: 0;
  width: 400%;
  left: 100%;

}

.drag_selection {
  position: absolute;
  background: $selection_square;
  border: 0px solid darken($selection_square, 10%);
  opacity: 0.2;
  pointer-events: none;
  z-index: $z_selection;
}


.cursor_select {
  cursor: default;
}

.cursor_drag {
  cursor: grab;
}

.cursor_dragging {
  cursor: grabbing;
}

.zoom_control {
  position: fixed;
  bottom: $unnnic-spacing-giant;
  left: 36px; 
  display: inline-flex;
  align-items: center;
  gap: $unnnic-spacing-xs;

  border-radius: $unnnic-border-radius-sm;
  background: #FFF;
  color: $unnnic-color-neutral-cloudy;

  box-shadow: $unnnic-shadow-level-near;

  .in, .out {
    display: flex;
    width: $unnnic-spacing-xl;
    height: $unnnic-spacing-xl;
    flex-direction: column;
    justify-content: center;
    align-items: center;
    gap: $unnnic-spacing-xs;
    background: #FFF;
    font-size: $unnnic-font-size-title-md;
    cursor: pointer;

    &:hover {
      background: $unnnic-color-neutral-light;
    }

    &:active {
      background: $unnnic-color-neutral-soft;
    }

    span {
      padding:  $unnnic-spacing-xs;
    }
  }

  .percentage {
    display: flex;
    width: 44px;
    padding: $unnnic-spacing-nano;
    justify-content: center;
    align-items: center;

    text-align: center;
    font-family: $unnnic-font-family-secondary;
    font-size: $unnnic-font-size-body-gt;
    line-height: 24px; /* 166.667% */

    span {
      font-size: 18px;
    }
  }
}