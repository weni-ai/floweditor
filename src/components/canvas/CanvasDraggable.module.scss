@import "src/variables.module.scss";

.draggable {
  position: absolute;

  -webkit-touch-callout: none;
  -webkit-user-select: none;
  -khtml-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;

  transition: left 0.2s cubic-bezier(0.175, 0.885, 0.32, 1.275),
    top 0.2s cubic-bezier(0.175, 0.885, 0.32, 1.275);
}

.selected {
  transition: none;
  z-index: $z_dragging;
}