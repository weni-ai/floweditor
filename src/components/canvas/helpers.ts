import { FlowPosition } from 'flowTypes';
import { CanvasPositions } from 'store/editor';
import {
  MAX_REFLOW_ATTEMPTS,
  NODE_SPACING,
  set,
  snapPositionToGrid,
  timeEnd,
  timeStart
} from 'utils';

const mutate = require('immutability-helper');

export const collides = (a: FlowPosition, b: FlowPosition, fudge: number) => {
  // don't bother with collision if we don't have full dimensions
  /* istanbul ignore next */
  if (!a.bottom || !b.bottom) {
    return false;
  }

  a.bottom += fudge;

  return !(b.left > a.right! || b.right! < a.left || b.top > a.bottom || b.bottom < a.top);
};

export const getDraggablesInBox = (
  positions: CanvasPositions,
  box: FlowPosition
): { [uuid: string]: FlowPosition } => {
  const collisions: any = {};
  for (const nodeUUID of Object.keys(positions)) {
    const position = positions[nodeUUID];
    if (collides(box, position, 0)) {
      collisions[nodeUUID] = position;
    }
  }
  return collisions;
};

interface DraggablePosition extends FlowPosition {
  uuid: string;
}

export const getOrderedDraggables = (positions: CanvasPositions): DraggablePosition[] => {
  const sorted: DraggablePosition[] = [];
  Object.keys(positions).forEach((uuid: string) => {
    sorted.push({ ...positions[uuid], uuid });
  });

  return sorted.sort((a: DraggablePosition, b: DraggablePosition) => {
    let diff = a.top - b.top;
    if (diff === 0) {
      diff = a.left - b.left;
    }
    return diff;
  });
};

/**
 * Gets the first collsion in the node map returning the original node,
 * the node it collides with and optionally an additional node it
 * collides with if inserting between two nodes
 * @param nodes
 */
const getFirstCollision = (
  positions: CanvasPositions,
  changed: string[],
  fudge: number
): DraggablePosition[] => {
  const sortedDraggables = getOrderedDraggables(positions);

  for (let i = 0; i < sortedDraggables.length; i++) {
    const current = sortedDraggables[i];

    if (i + 1 < sortedDraggables.length) {
      for (let j = i + 1; j < sortedDraggables.length; j++) {
        const other = sortedDraggables[j];

        if (collides(current, other, fudge)) {
          // if the next node collides too, include it
          // to deal with inserting between two closely
          // positioned nodes
          if (j + 1 < sortedDraggables.length) {
            const cascaded = sortedDraggables[j + 1];
            if (collides(other, cascaded, fudge)) {
              return [current, other, cascaded];
            }
          }

          if (
            !!changed.find((uuid: string) => other.uuid === uuid) &&
            !!!changed.find((uuid: string) => current.uuid === uuid)
          ) {
            return [other, current];
          }

          return [current, other];
        }
      }
    }
  }
  return [];
};

const setTop = (position: FlowPosition, newTop: number) => {
  return {
    top: newTop,
    left: position.left,
    bottom: newTop + (position.bottom! - position.top),
    right: position.right
  };
};

/**
 * Reflow positions to account for any collisions
 * @param positions
 */
export const reflow = (
  positions: CanvasPositions,
  fudge: number,
  startingNodeUuid: string
): { positions: CanvasPositions; changed: string[] } => {
  let newPositions = positions;
  const changed: string[] = [];

  // if for some reason we can't reflow, don't blow up
  let attempts = 0;

  timeStart('reflow');

  let collision = getFirstCollision(positions, changed, fudge);
  while (collision.length > 0 && attempts < MAX_REFLOW_ATTEMPTS) {
    attempts++;
    if (collision.length) {
      const [top, bottom, cascade] = collision;
      if (bottom.uuid !== startingNodeUuid) {
        newPositions = mutate(newPositions, {
          [bottom.uuid]: set(setTop(bottom, top.bottom! + NODE_SPACING))
        });
        changed.push(bottom.uuid);
      } else {
        newPositions = mutate(newPositions, {
          [top.uuid]: set(setTop(top, bottom.top! + NODE_SPACING))
        });
        changed.push(top.uuid);
      }

      if (cascade) {
        // start with the top of the bottom node
        let cascadeTop = top.bottom! + NODE_SPACING;

        // and add its height
        cascadeTop += bottom.bottom! - bottom.top;
        newPositions = mutate(newPositions, {
          [cascade.uuid]: set(setTop(cascade, cascadeTop))
        });

        changed.push(cascade.uuid);
      }
    }

    collision = getFirstCollision(newPositions, changed, fudge);
  }

  timeEnd('reflow');

  return { positions: newPositions, changed };
};
