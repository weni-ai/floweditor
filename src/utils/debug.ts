import { FlowStoreProps } from 'components/flow/Flow';
import { DebugState } from 'store/editor';

import mutate from 'immutability-helper';

/* istanbul ignore next -- @preserve */
export default class Debug {
  private props: FlowStoreProps;
  private state: DebugState;

  constructor(props: FlowStoreProps, initial: DebugState) {
    this.props = props;
    this.state = initial || { showUUIDs: false };
  }

  public showUUIDs(): DebugState {
    const updated = mutate(this.state, { $merge: { showUUIDs: true } });
    this.props.mergeEditorState({ debug: updated });
    return updated;
  }
}
