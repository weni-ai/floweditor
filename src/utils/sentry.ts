import * as Sentry from '@sentry/react';

const initSentry = (dsn: string) => {
  Sentry.init({
    dsn,
    integrations: [
      Sentry.browserTracingIntegration(),
      Sentry.httpClientIntegration({
        failedRequestStatusCodes: [[400, 499], [500, 599]],
      }),
      Sentry.replayIntegration(),
      Sentry.replayCanvasIntegration(),
    ],
    tracesSampleRate: 1.0,
    replaysSessionSampleRate: 0.1,
    replaysOnErrorSampleRate: 1.0,
  });
};

export default initSentry;
