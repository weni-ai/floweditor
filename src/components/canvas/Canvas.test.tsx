import { Canvas, CANVAS_PADDING, CanvasProps } from 'components/canvas/Canvas';
import { CanvasDraggableProps } from 'components/canvas/CanvasDraggable';
import React from 'react';
import { fireEvent, render } from 'test/utils';
import { createUUID } from 'utils';

jest.mock('components/sidebar/Sidebar', () => () => 'Mocked Sidebar');

const baseProps: CanvasProps = {
  uuid: createUUID(),
  draggingNew: false,
  dragActive: false,
  onDragging: jest.fn(),
  onUpdatePositions: jest.fn(),
  mergeEditorState: jest.fn(),
  onRemoveNodes: jest.fn(),
  onDoubleClick: jest.fn(),
  onLoaded: jest.fn(),
  draggables: [],
  newDragElement: <div></div>,
  mutable: true
};

describe(Canvas.name, () => {
  it('render default', () => {
    const { baseElement } = render(<Canvas {...baseProps} />);
    expect(baseElement).toMatchSnapshot();
  });

  it('initializes the height to the lowest draggable', () => {
    const lowest: CanvasDraggableProps = {
      elementCreator: jest.fn(),
      uuid: createUUID(),
      position: { top: 1200, left: 100, bottom: 1290, right: 300 },
      idx: 0
    };
    const { baseElement } = render(<Canvas {...baseProps} draggables={[lowest]} />);
    expect(baseElement).toMatchSnapshot();
  });

  it('adjusts the height when updating dimensions', () => {
    const uuid = createUUID();
    const lowest: CanvasDraggableProps = {
      elementCreator: jest.fn(),
      uuid,
      position: { top: 1200, left: 100, right: 200, bottom: 1400 },
      idx: 0
    };

    const { baseElement, getByTestId } = render(<Canvas {...baseProps} draggables={[lowest]} />);
    expect(baseElement).toMatchSnapshot();
  });

  it('reflows collisions', () => {
    jest.useFakeTimers();

    const first: CanvasDraggableProps = {
      elementCreator: jest.fn(),
      uuid: createUUID(),
      position: { top: 100, bottom: 200, left: 100, right: 200 },
      idx: 0
    };

    const second: CanvasDraggableProps = {
      elementCreator: jest.fn(),
      uuid: createUUID(),
      position: { top: 150, left: 100, bottom: 250, right: 200 },
      idx: 0
    };

    const onDragging = jest.fn();

    const { getByTestId } = render(
      <Canvas {...baseProps} draggables={[first, second]} onDragging={onDragging} />
    );

    // trigger reflow by simulating a drag event
    fireEvent.mouseDown(getByTestId('draggable_' + first.uuid));
    fireEvent.mouseUp(getByTestId('draggable_' + first.uuid));
    jest.runAllTimers();

    expect(onDragging).toMatchCallSnapshot();
  });
});
