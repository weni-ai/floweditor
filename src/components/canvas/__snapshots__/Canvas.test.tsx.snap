// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Canvas adjusts the height when updating dimensions 1`] = `
<body>
  <div>
    <div
      class="canvas_container cursor_dragging"
      id="canvas-container"
    >
      <div
        class="canvas_background"
      >
        <div
          class="canvas"
          data-testid="canvas"
          id="canvas"
          tabindex="0"
        >
          <div
            class="panzoom"
            id="panzoom"
          >
            <div />
            <div
              class="draggable"
              data-testid="draggable_132de855-4042-4dc1-a18f-cc2e6a8f790a"
              style="left: 100px; top: 1200px;"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`Canvas initializes the height to the lowest draggable 1`] = `
<body>
  <div>
    <div
      class="canvas_container cursor_dragging"
      id="canvas-container"
    >
      <div
        class="canvas_background"
      >
        <div
          class="canvas"
          data-testid="canvas"
          id="canvas"
          tabindex="0"
        >
          <div
            class="panzoom"
            id="panzoom"
          >
            <div />
            <div
              class="draggable"
              data-testid="draggable_1e47a1e1-3c67-4df5-adf1-da542c789adb"
              style="left: 100px; top: 1200px;"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`Canvas reflows collisions 1`] = `
Array [
  Array [
    "42ebf1d0-a43e-4ba6-9577-c863b20c88c4",
  ],
]
`;

exports[`Canvas render default 1`] = `
<body>
  <div>
    <div
      class="canvas_container cursor_dragging"
      id="canvas-container"
    >
      <div
        class="canvas_background"
      >
        <div
          class="canvas"
          data-testid="canvas"
          id="canvas"
          tabindex="0"
        >
          <div
            class="panzoom"
            id="panzoom"
          >
            <div />
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;
